core/src/main/java/com/sequenceiq/cloudbreak/service/multiaz/MultiAzCalculatorService.java

```diff
@@ -248,4 +254,123 @@ private boolean isPreferredSubnetsSpecifiedForScaling(NetworkScaleDetails networ
         return networkScaleDetails != null
                 && CollectionUtils.isNotEmpty(networkScaleDetails.getPreferredSubnetIds());
     }
+
+    public Set<GroupPlacement> prepareGroupPlacementSet(Map<String, String> subnetAzPairs, Set<InstanceGroup> instanceGroups) {
+        Set<String> hints = getHints(instanceGroups);
+        Set<String> availabilityZones = getAvailabilityZones(subnetAzPairs);
+
+        if (hints.size() <  MINIMUM_NUMBER_OF_HINTS || availabilityZones.size() < hints.size()) {
+            return Collections.emptySet();
+        }
+
+        Set<GroupPlacement> groupPlacements = new HashSet<>();
+        Iterator<String> availabilityZonesIt = availabilityZones.iterator();
+        hints.forEach(hint -> {
+            String availabilityZone = availabilityZonesIt.next();
+            GroupPlacement groupPlacement = new GroupPlacement(hint, availabilityZone);
+            groupPlacements.add(groupPlacement);
+
+            subnetAzPairs
+                    .entrySet()
+                    .stream()
+                    .filter(entry -> entry.getValue().equals(availabilityZone))
+                    .map(Entry::getKey)
+                    .forEach(groupPlacement::addSubnet);
+        });
+        return groupPlacements;
+    }
+
+    private Set<String> getAvailabilityZones(Map<String, String> subnetAzPairs) {
+        return subnetAzPairs
+                .values()
+                .stream()
+                .collect(Collectors.toSet());
+    }
+
+    private Set<String> getHints(Set<InstanceGroup> instanceGroups) {
+        return instanceGroups
+                .stream()
+                .filter(ig -> Objects.nonNull(ig.getHints()))
+                .map(InstanceGroup::getHints)
+                .flatMap(Set::stream)
+                .collect(Collectors.toSet());
+    }
+
+    public void calculate(Map<String, String> subnetAzPairs, InstanceGroup instanceGroup, Set<GroupPlacement> groupPlacements) {
+        if (isGroupPlacement(groupPlacements)) {
+            calculateByHints(instanceGroup, groupPlacements);
+        } else {
+            calculateByRoundRobin(subnetAzPairs, instanceGroup);
+        }
+    }
+
+    private boolean isGroupPlacement(Set<GroupPlacement> groupPlacements) {
+        return !groupPlacements.isEmpty();
+    }
+
+    private void calculateByHints(InstanceGroup instanceGroup, Set<GroupPlacement> groupPlacementList) {
+        LOGGER.info("Calculate the subnet by group");
+        List<InstanceMetadataView> instanceMetadataViews = new ArrayList<>(instanceGroup.getNotDeletedAndNotZombieInstanceMetaDataSet());
+        groupPlacementList.forEach(groupPlacement -> collectCurrentSubnetUsage(instanceMetadataViews, groupPlacement.getSubnetUsage()));
```

Comment:
Why are we considering all subnets for environment here? There can be subnets for instance group and those should be used if available. Please refer to collectSubnetIds() method in this class.